### PR TITLE
MLflow autologging

### PR DESCRIPTION
**Pull request by wd60622**
## Description

General autologging for PyMC models

```python
from pymc_marketing.mlflow import autolog

autolog()

# Any PyMC workflow / model

import pymc as pm

coords = ...
with pm.Model(coords=coords) as model:
    ...
```

Some additional params, metrics, and artifacts with MMM class

```python
from pymc_marketing.mlflow import autolog
from pymc_marketing.mmm import MMM, GeometricAdstock, LogisticSaturation

autolog()
mmm = MMM(...)
mmm.fit(X, y)
```

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by wd60622** [comment]
Some items: 
- should we install blackjax, nutpie, and numpyro for these tests?
- Installation of graphviz in the runs for the graph artifact
- Unpatching the `sample` and `fit` so that the rest of the tests are not affected

File: pymc_marketing/mlflow.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    raise ImportError(msg)
+
+from mlflow.utils.autologging_utils import autologging_integration
+
+from pymc_marketing.mmm import MMM
+
+FLAVOR_NAME = "pymc"
+
+
+def save_arviz_summary(idata: az.InferenceData, path: str | Path, var_names) -> None:
```
Comment (juanitorduz) [inline pymc_marketing/mlflow.py:36]: missing type hint in var-names

Comment (juanitorduz) [reply pymc_marketing/mlflow.py:36]: and add docstrings

File: pymc_marketing/mlflow.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    data = mlflow.data.from_numpy(features=features, targets=targets)
+    mlflow.log_input(data, context="sample")
+
+
+def save_model_graph(model: Model, path: str | Path) -> None:
+    try:
+        graph = pm.model_to_graphviz(model)
+    except ImportError:
+        return None
```
Comment (juanitorduz) [inline pymc_marketing/mlflow.py:63]: Shall we raise here something more informative so that this does not pass quietly?

**Review by juanitorduz — COMMENTED** [review]
This looks amazing! We should have an issue having a notebook or some docs on how to set up this locally and how this works for people new to mlflow.

**Comment by juanitorduz** [comment]
> Some items:
> 
> * should we install blackjax, nutpie, and numpyro for these tests?

Yes, we are installing them for the docs anyway (for the other samplers notebook)

> * Installation of graphviz in the runs for the graph artifact

What do you mean?

> * Unpatching the `sample` and `fit` so that the rest of the tests are not affected

Probably ... Can we just provide a decorator so that is handled by the user, say for example

```python
mmm = MMM ( ... )

@log_mlflow
mmm.fit(X, y)
```

File: pymc_marketing/mlflow.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    pymc_marketing.mlflow.autolog(log_mmm=True)
+
+    # Usual PyMC-Marketing model code
+
+    data_url = "https://raw.githubusercontent.com/pymc-labs/pymc-marketing/main/data/mmm_example.csv"
+    data = pd.read_csv(data_url, parse_dates=["date_week"])
+
+    X = data.drop("y",axis=1)
+    y = data["y"]
+    mmm.fit(X,y)
```
Comment (wd60622) [inline pymc_marketing/mlflow.py:75]:
```suggestion
```

File: pymc_marketing/mlflow.py (hunk 2 above)
Comment (wd60622) [reply pymc_marketing/mlflow.py:199]: I have a test for the behavior. If the import doesn't work then it won't be logged via MLflow. 
Should it be silent? I don't think it should raise

**Comment by wd60622** [comment]
Added some more examples and documentation. 

The pytest fixture will manage unpatching the functions after the module is tested. That should handle the failures that were happening previously.

Let me know your thoughts @juanitorduz

**Comment by wd60622** [comment]
Failure due to `model.data_vars` not being an attribute in the PyMC=5.13.0

**Comment by wd60622** [comment]
Just handful of conditionals that don't have coverage now. Check those out here: https://app.codecov.io/gh/pymc-labs/pymc-marketing/pull/921?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=pymc-labs

Other than that, should be good to go

File: pymc_marketing/mlflow.py (hunk 2 above)
Comment (juanitorduz) [reply pymc_marketing/mlflow.py:218]: I think if the import fails and the user actively wants to logs and something fails then we should at least get a warning

**Comment by juanitorduz** [comment]
> Just handful of conditionals that don't have coverage now. Can add in some additional checks if you'd like Check those out here: https://app.codecov.io/gh/pymc-labs/pymc-marketing/pull/921?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=pymc-labs
> 
> Other than that, should be good to go

I think you can add `# pragma: no cover` to a line so that codecov does not take into account

File: pymc_marketing/mlflow.py (hunk 2 above)
Comment (wd60622) [reply pymc_marketing/mlflow.py]: Added a logging info level. That feel sufficient?

File: pymc_marketing/mlflow.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+    idata : az.InferenceData
+        The InferenceData object returned by the sampling method.
+    path : str | Path
+        The path to save the summary as HTML.
+    var_names : list[str], optional
+        The names of the variables to include in the summary. Default is
+        all the variables in the InferenceData object.
+
+    """
+    df_summary = az.summary(idata, var_names=var_names)
```
Comment (juanitorduz) [inline pymc_marketing/mlflow.py:146]: can we allow `kwargs` for az.summary?

File: pymc_marketing/mlflow.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    df_summary.to_html(path)
+    mlflow.log_artifact(str(path))
+    os.remove(path)
+
+
+def _backwards_compatiable_data_vars(model: Model) -> list[TensorVariable]:
+    # TODO: Remove with PyMC update
```
Comment (juanitorduz) [inline pymc_marketing/mlflow.py:153]: can we create an issue about this?

**Review by juanitorduz — APPROVED** [review]
Hey! This looks super nice! I think we won't get it 100% right the first time so I suggest we ship it and get feedback . Do we want to add a warning saying that this is an experimental feature?

I left a couple of small suggestions. Let me know what you thing about it.

Maybe this should be the highlight of the next release?

File: docs/source/api/index.md [hunk 6]
```diff
@@ -13,4 +13,5 @@
   model_config
   model_builder
   prior
+  mlflow
```
[the remark this answers is not in the document]
Comment (wd60622) [reply docs/source/api/index.md:16]: This is just the name of the module pymc_marketing.mlflow and not the dependency

File: pyproject.toml [hunk 7]
```diff
@@ -65,13 +65,17 @@ docs = [
...
+    "nutpie",
+    "numpyro",
     "graphviz==0.20.1",
     "lifetimes==0.11.3",
     "papermill",
     "preliz==0.8.0",
     "pytest-cov==3.0.0",
     "pytest-mock==3.14.0",
     "pytest==7.0.1",
+    "mlflow>=2.0.0",
```
[the remark this answers is not in the document]
Comment (wd60622) [reply pyproject.toml:78]: This is just for the test dependencies which tests with SQL connection as well. If the import doesn't work then I have an error with suggestion to install. It is not a dependency of the package

File: pymc_marketing/mlflow.py (hunk 5 above)
Comment (wd60622) [reply pymc_marketing/mlflow.py:164]: done!

**Comment by wd60622** [comment]
Any idea why the new module doesnt show up? It works locally

https://pymc-marketing--921.org.readthedocs.build/en/921/api/index.html

**Comment by juanitorduz** [comment]
> Any idea why the new module doesnt show up? It works locally
> 
> https://pymc-marketing--921.org.readthedocs.build/en/921/api/index.html

No idea :(

**Comment by wd60622** [comment]
It should. Anything seem missing? 
The docs were failing at first, maybe that has something to do with it

**Comment by juanitorduz** [comment]
Maybe we can simply merge and see? If it does not work we open a PR just to try to fix this

**Pull request merged**